packages/telemetry/src/context-aware-slog.js

```diff
@@ -0,0 +1,475 @@
+       */
+      // @ts-expect-error
+      const row = databaseInstance
+        .prepare(
+          `SELECT * FROM '${DATABASE_TABLE_NAME}' WHERE key = '${DATABASE_TRIGGER_CONTEXT_KEY_NAME}'`,
+        )
+        .get();
+
+      delete row?.key;
+
+      return row || null;
+    },
+  };
+};
+
+const stringify = data =>
+  JSON.stringify(data, (_, value) =>
+    typeof value === 'bigint' ? Number(value) : value,
+  );
+
+/**
+ *
+ * @param {{env: typeof process.env}} options
+ */
+export const makeSlogSender = async options => {
+  const { CHAIN_ID, OTEL_EXPORTER_OTLP_ENDPOINT } = options.env;
+  if (!OTEL_EXPORTER_OTLP_ENDPOINT)
+    return console.warn(
+      'Ignoring invocation of slogger "context-aware-slog" without the presence of "OTEL_EXPORTER_OTLP_ENDPOINT" envrionment variable',
+    );
+
+  const loggerProvider = new LoggerProvider({
+    resource: new Resource(getResourceAttributes(options)),
+  });
+  loggerProvider.addLogRecordProcessor(
+    new SimpleLogRecordProcessor(new OTLPLogExporter({ keepAlive: true })),
+  );
+
+  logs.setGlobalLoggerProvider(loggerProvider);
+  const logger = logs.getLogger('default');
+
+  const db = await getDatabaseInstance();
+
+  /** @type Array<Context | null> */
+  let [
+    blockContext,
+    crankContext,
+    initContext,
+    lastPersistedTriggerContext,
+    replayContext,
+    triggerContext,
+  ] = [null, null, null, null, null, null];
+
+  /**
+   * @param {Context} context
+   */
+  const persistContext = context => {
+    lastPersistedTriggerContext = context;
+    db.persistContext(context);
+  };
+
+  const restoreContext = () => {
+    if (lastPersistedTriggerContext) return lastPersistedTriggerContext;
+    return db.restoreContext();
+  };
+
+  /**
+   * @param {Slog} slog
+   */
+  const slogSender = ({ monotime, time: timestamp, ...body }) => {
+    const finalBody = { ...body };
+
+    /** @type {LogAttributes} */
+    let logAttributes = {
+      'chain-id': String(CHAIN_ID),
+      'process.uptime': monotime,
+      timestamp,
+    };
+
+    /**
+     * Add any before report operations here
+     * like setting context data
+     */
+    switch (body.type) {
+      case SLOG_TYPES.CONSOLE: {
+        delete finalBody.crankNum;
+        delete finalBody.deliveryNum;
+
+        break;
+      }
+      case SLOG_TYPES.CLIST: {
+        assert(!!crankContext);
+        crankContext['crank.vatID'] = finalBody.vatID;
+        break;
+      }
+      case SLOG_TYPES.CRANK.START: {
+        crankContext = {
+          'crank.num': finalBody.crankNum,
+          'crank.type': finalBody.crankType,
+        };
+        break;
+      }
+      case SLOG_TYPES.DELIVER: {
+        if (replayContext) {
+          assert(finalBody.replay);
+          replayContext = {
+            ...replayContext,
+            'crank.deliveryNum': finalBody.deliveryNum,
+            'crank.vatID': finalBody.vatID,
+          };
+        } else {
+          assert(!!crankContext);
```

Comment:
We probably should also assert that the event is not replaying

```suggestion
          assert(!!crankContext);
          assert(!finalBody.replay);
```